### ISSUE TITLE
new and safeNew should return a vector of size 0

### ISSUE THREAD

**Issue by OlivierSohn** [AndrasKovacs/dynamic-mvector#6]
Today, new and safeNew return a vector of length equal to its capacity, but the notion of length is the count of valid elements so it should be zero in this case because elements are uninitialized.

**Comment by AndrasKovacs** [comment]
The notion of length should be the size of the range where you can read/write without throwing exception. It's currently the same behavior as `new` in [Data.Vector.Mutable](https://hackage.haskell.org/package/vector-0.12.0.1/docs/Data-Vector-Mutable.html).

**Comment by OlivierSohn** [comment]
But why does the doc of new say "The elements are uninitialized and throw error upon accessing." ?
It should rather say "The elements are uninitialized". Or, we could pass an element value to initialize the elements to this value. I find it confusiong that you can easily create a vector whose elements are both accessible and uninitialized. In c++, when you create a new vector of size n, you have two kinds of constructors : thoses were you pass also the value to which all elements should be initialized, and those who say the value will be default-initialized. So you know that they will have the default value. But here, since there is no notion of default value in Haskell, the API should be changed I think.

**Comment by AndrasKovacs** [comment]
Yeah, if you want a new vector initialized with an element, it would be bad to first initialize it with undefined and then copy over again the intended element. It makes sense to make `new` accept an initial element, and not follow  Data.Vector.Mutable API in this matter.

**Comment by OlivierSohn** [comment]
ok, so we could add a `Maybe` value to `new` and `unsafeNew`:

if we want a n-size vector we do

```haskell
v <- new n $ Just initialValue
```

and if we want a 0-size vector of n capacity:

```haskell
v <- new 0 Nothing
reserve v n
```
EDIT : the problem with the `Maybe` solution is that we could write

```haskell
v <- new n Nothing
```
and then we have again a vector of uninitialized Elements. So maybe the solution is to split the functions in 2:

if we want a n-size vector we do

```haskell
v <- newWith n initialValue
```

and if we want a 0-size vector of n capacity:

```haskell
v <- new
reserve v n
```

EDIT 2: Note this is a breaking change, so if it is implemented the version should be properly incremented on hackage (I think it's the 2nd digit that should be incremented?)

**Comment by AndrasKovacs** [comment]
I think we could have `new :: PrimMonad m => Int -> a -> m (MVector (PrimState m) a)` which doesn't reserve any extra space and initializes to `a`, and `newWithReserve :: PrimMonad m => Int -> a -> Int -> m (MVector (PrimState m) a)` which does reserve.

I see now that the `new` function `Data.Vector.Mutable` is kind of crap, because it always initializes with an undefined element. Instead, `newArray#` from [here](https://hackage.haskell.org/package/ghc-prim-0.5.0.0/docs/GHC-Prim.html#t:MutableArray-35-) should be used for initialization.

**Comment by OlivierSohn** [comment]
For newWithReserve you propose:

```haskell
newWithReserve :: PrimMonad m => Int -> a -> Int -> m (MVector (PrimState m) a)
```

The a passed as argument, is it just to give the type, and we ignore the value? 

And why are there 2 Ints? I think one would suffice?

**Comment by AndrasKovacs** [comment]
The first `Int` is the number of initialized elements, the second could be additional reserved space, but other configurations of arguments may work as well. The idea is that since GHC does not actually have uninitialized boxed array creation, we always want to provide initial elements. Users can just choose to initialize  with `undefined` etc.

**Comment by OlivierSohn** [comment]
I find it a bit confusing that at the same time we initialize part of the vector with the value and reserve the rest. I don't see a use case for that.

What do you think of
```haskell
newWithReserve :: PrimMonad m => Int -> m (MVector (PrimState m) a)
```

where we just reserve, and do no initialization?

**Comment by AndrasKovacs** [comment]
Well, if you do want to both initialize and reserve, then (since GHC does not have uninitialized array creation), you first do `newWithReserve`, which necessarily allocates and copies the whole array over with some `undefined` element, then you copy over once again the parts which you want to initialize to some `a`. If there's a combined function, we can just initialize everything once to `a`. 

The issue here though is that we would also have to initialize the "reserved" parts to `a`, which can result in silently returning valid `a`-s when unsafely accessing reserved indices, instead of throwing exceptions. I don't know how problematic this is; I guess since unsafe indexing can already do any kind of undefined behavior by indexing out of reserved space even, this is okay. If someone wants guaranteed exceptions on out-of-bound indexing, they should just use safe indexing.

**Comment by OlivierSohn** [comment]
Yes, that's why I think `new` covers the case where you initialize the vector values, and `newWithReserve` should just reserve, not initialize. If you look at the c++ vector API there is no mixed way to both initialize some and reserve, it's also more difficult to grasp as a user of the API I think.
